### DISCUSSION TITLE
UpdateMachineAuthCallback is not triggered

### DISCUSSION THREAD

**Discussion by ExistentialDysfunction** [SteamRE/SteamKit#1258] [General]
Hello, I have added a functionality to fetch steam guard code from the email, but after successful authorization there is no UpdateMachineAuthCallback event triggered, so there is no sentry file generated for some reason. Any ideas what might be the issue?

**Comment by zxcvqwerasdf** [comment]
Becase it's deprecated now. It's not an issue.

**Reply by HexagonAV** [reply]
As I understand it, sentry was needed so as not to enter Steam Guard every time? What replacement for senty is being used now?

**Reply by HexagonAV** [reply]
I tried to see what happened during authorization and noticed that the server was sending refresh_token and access_token. Are they used instead of sentry?

**Reply by LossyDragon** [reply]
Yes, steam no longer send sentry files in favor of the new Login Authentication system. 

Samples 1a and 1b show how to use the new authentication. Note that `IsPersistentSession` should be true if you are looking to keep the same tokens to subsequent logins. Then your account name and RefreshToken should be saved to pass into LogOn using some logic to check if you have them stored while bypassing the authentication part.

**Reply by ExistentialDysfunction** [reply]
Does the token expire at some point?

**Reply by LossyDragon** [reply]
I believe they can either by: temporary session, revoked via mobile app, or if steam expires the session for whatever reason.

Sample 1a has a function to parse the JWT response, I do recall it having some expiry info in it. Maybe that can help answer any other questions about session expiration. I can't recall what all is has in the response since I haven't parsed my tokens since the PR to handle the new auth came out.

**Reply by ExistentialDysfunction** [reply]
Okay, so I have implemented an auto reconnect logic and it works fine. JWT tokens are stored for all bots and upon disconnect they reconnect automatically within 5 seconds if no other critical errors, but after they start reconnecting I always get this error multiple times before managing to log back in successfully:

Unable to logon to Steam: 0 / 1
 

Why is that happening? How to better reconnect all bots after they start disconnecting in about 4-6 hours or so? Some of the bots can reconnect after first retry, others get this same error multiple times and I can't figure out why.

**Reply by LossyDragon** [reply]
> Why is that happening?

Was this happening before? Are you behind a proxy or something that could make a connection to the steam servers unstable?

Once you call `SteamUser.LogOn()` you're pretty much using the same features as before using the old login method or new. 

If your bots are all using the same account you may need to try adding `SteamUser.LoginID`?

I wouldn't be able to help much here since I only use SteamKit as reference for the Java port. Hopefully somebody can chime in with better solutions for you.

**Reply by ExistentialDysfunction** [reply]
I fixed this problem, but now I am getting rate limited After logging in 250-300 bots. There is a piped connection login of 3 connections concurrently. After every successful logOn there is a 500ms delay as well. Dunno how to properly handle the rate limit error now. Any idea what are the exact rate limits?

**Reply by STRATZ-Ken** [reply]
Can you guys explain what 1a and 1b is? Am I missing something?

**Reply by LossyDragon** [reply]
> Can you guys explain what 1a and 1b is? Am I missing something?

Two of the sample projects.
[Sample 1a](https://github.com/SteamRE/SteamKit/tree/master/Samples/1a.Authentication)
[Sample 1b](https://github.com/SteamRE/SteamKit/tree/master/Samples/1b.QrCodeAuthentication)

**Reply by STRATZ-Ken** [reply]
Yeah thanks Lossy, I just found them. I saw 1 and 2, but somehow I missed 1a.

**Reply by STRATZ-Ken** [reply]
Is there anyway to get the account name from `public Task<string> GetEmailCodeAsync(string email, bool previousCodeWasIncorrect)` ?

I see you send the email that is responded, but if I have a few accounts, how do I know which code to look for?

**Reply by STRATZ-Ken** [reply]
I was able to modify the core code to get this through.  

However, I have a question.  Once I get the request for a 2FA, should I still get disconnected and then relog with the 2FA?  Or should the new system polling be responsible for keeping that connection open, but then sending the 2FA when it comes in?  Emails can take up to 60 seconds to arrive.  I keep getting OnDisconnected being called before I can get the 2FA from the email service.

Also, should RefreshTokens remain active for long periods?  When I saw the token `exp` it says May of next year, but when I try to log in again a few days after getting the token, it says AccessDenied.

**Reply by STRATZ-Ken** [reply]
Wanted to include an example.  I believe its some type of race condition.

![image](https://github.com/SteamRE/SteamKit/assets/24981405/2351353d-7d6e-4eeb-80cc-a159bc4e875d)

In the attached image, you can see I am attempting to connect to Steam with a Refresh Token.  (No Idea why, but it expired and is invalid, if anyone has information on this, would love to know, Persistent = `true` already).  Once I notice that the 2FA is required, it disconnects and now is going through the login process again.  

I log in again, however this time I got 2FA.  (Which I should).  Now, its currently `authSession.PollingWaitForResultAsync` but again while I am waiting it disconnects.  By the time I get the 2FA in, the connection is gone.  Do I reuse the old 2FA token here or do I wait for the new one and hope the connect stays until its completed.  

Lastly, it then gets the 2FA, and attempts to send the 2FA to Steam, but the connection is already dropped and thus we get `Attempting to send while not connected` error.

Thanks,

**Reply by Christian-Stieber** [reply]
> Once I get the request for a 2FA, should I still get disconnected and then relog with the 2FA? 

No, you won't get disconnected... I think. It's been a while since I worked on this, but a quick test with an EMail-guarded account and a look into the log suggests there's no disconnect happening anymore.
I'm still having some problem with EMail-guard code handling, but that's likely unrelated to the question.

For me, the basic flow looks like this:
* I'm sending CAuthentication_BeginAuthSessionViaCredentials_Request
* I'm receiving CAuthentication_BeginAuthSessionViaCredentials_Response, telling me the account has guard-email
* I receive the EMail, and copy-paste the code into the bot
* I send CAuthentication_UpdateAuthSessionWithSteamGuardCode_Request with the code
* I receive CAuthentication_UpdateAuthSessionWithSteamGuardCode_Response
* I send CAuthentication_PollAuthSessionStatus_Request
* I receive CAuthentication_PollAuthSessionStatus_Response, giving me my refresh token (which I'm storing for future logins)
* I perform the normal CMsgClientLogon

> Also, should RefreshTokens remain active for long periods?

Yes. I expect them to expire eventually, but haven't seen it happen yet.

> When I saw the token exp it says May of next year

Oh... so I guess SteamKit has code to decipher these tokens? Good to know, I want that too :-)
...
Ok, found ParseJsonWebToken() :-)

**Reply by STRATZ-Ken** [reply]
The issue with the expiration is, if you call LogOff at any point, all your tokens are now invalid.   However, if you call a Disconnect, you can reconnect using the same token.  Thanks, all is working now. Though, I completely removed the Polling and made my own method as it was not dynamic enough to support everything I needed.

**Reply by xPaw** [reply]
To get account in `GetEmailCodeAsync`, copy `UserConsoleAuthenticator` implementation, and change it to pass in whatever data you want, so when you do `Authenticator = new UserConsoleAuthenticator()`, you can instead do `Authenticator = new MyAuthenticator(accountName)` for example.

The IAuthenticator is very simple, and specifically was made for things like this.

To handle email 2FA, you have to store GuardData that comes from PollingWaitForResultAsync, and then pass it into AuthSessionDetails, this works similarly to old sentry files.

See https://github.com/SteamRE/SteamKit/commit/e3950f948524d8517e82692f4a233e67325367f7

**Reply by skylayer** [reply]
I'm working through sample 1a and hit a snag with the `NewGuardData` from Steam's 2FA. After successfully logging in and approving the login on my phone, I expected to receive usable guard data for future logins. However, the `NewGuardData` appears empty. Here's the relevant code snippet and the exact output:

```csharp
if (pollResponse.NewGuardData != null) {
    previouslyStoredGuardData = pollResponse.NewGuardData;
    Console.WriteLine($"Received new guard data, store this for future logons: {previouslyStoredGuardData}");
    Console.WriteLine($"Length of guard data: {previouslyStoredGuardData.Length}");
}
```

Output:
```
Connecting to Steam...
Connected to Steam! Logging in '[username]'...
STEAM GUARD! Use the Steam Mobile App to confirm your sign in...
Received new guard data, store this for future logons: 
Length of guard data: 0
```

Has anyone else experienced this? Looking for insights or solutions.

**Comment by ExistentialDysfunction** [comment]
So, it is currently impossible to receive and update the sentry file?
